File: dotnet/Goldy/SurfaceExchange.cs
using Goldy.Native;

namespace Goldy;

/// <summary>
/// Window-surface exchange for present-on-scheme.
/// </summary>
public sealed class SurfaceExchange : IDisposable
{
    internal readonly nint Handle;
    private bool _disposed;

    private SurfaceExchange(nint handle) => Handle = handle;

    public static SurfaceExchange CreateWin32(Context ctx, nint hwnd, uint depth = 3)
    {
        ctx.ThrowIfDisposed();
        var handle = NativeMethods.SurfaceExchangeCreateWin32(ctx.Handle, hwnd, depth);
        if (handle == nint.Zero)
            throw GoldyException.FromLastError("SurfaceExchange creation");
        return new SurfaceExchange(handle);
    }

    public static SurfaceExchange CreateAppKit(Context ctx, nint nsView, uint depth = 3)
    {
        ctx.ThrowIfDisposed();
        var handle = NativeMethods.SurfaceExchangeCreateAppKit(ctx.Handle, nsView, depth);
        if (handle == nint.Zero)
            throw GoldyException.FromLastError("SurfaceExchange creation");
        return new SurfaceExchange(handle);
    }

    public static SurfaceExchange CreateWayland(Context ctx, nint display, nint surface, uint depth = 3)
    {
        ctx.ThrowIfDisposed();
        var handle = NativeMethods.SurfaceExchangeCreateWayland(ctx.Handle, display, surface, depth);
        if (handle == nint.Zero)
            throw GoldyException.FromLastError("SurfaceExchange creation");
        return new SurfaceExchange(handle);
    }

    public uint Width => NativeMethods.SurfaceExchangeWidth(Handle);

    public uint Height => NativeMethods.SurfaceExchangeHeight(Handle);

    public TextureFormat Format => NativeMethods.SurfaceExchangeFormat(Handle);

    public ulong Generation => NativeMethods.SurfaceExchangeGeneration(Handle);

    public void Resize(uint width, uint height)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        if (width == 0 || height == 0)
            return;
        var result = NativeMethods.SurfaceExchangeResize(Handle, width, height);
        if (result != GoldyResult.Ok)
            throw GoldyException.FromLastError("SurfaceExchange resize");
    }

    public PresentLease Lease()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        var lease = NativeMethods.SurfaceExchangeLease(Handle);
        if (lease == nint.Zero)
            throw GoldyException.FromLastError("SurfaceExchange lease");
        return new PresentLease(lease);
    }

    public Transaction BindRenderTarget(Scheme scheme, SchemeRenderTargetLease source)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(scheme);
        ArgumentNullException.ThrowIfNull(source);
        var tx = NativeMethods.SurfaceExchangeBindRenderTarget(Handle, scheme.Handle, source.Handle);
        if (tx == nint.Zero)
            throw GoldyException.FromLastError("SurfaceExchange bind_render_target");
        return new Transaction(tx);
    }

    public Transaction Bind(Scheme scheme, Texture source)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(scheme);
        ArgumentNullException.ThrowIfNull(source);
        var tx = NativeMethods.SurfaceExchangeBind(Handle, scheme.Handle, source.Handle);
        if (tx == nint.Zero)
            throw GoldyException.FromLastError("SurfaceExchange bind");
        return new Transaction(tx);
    }

    public (PresentLease Lease, Transaction Transaction) BindDestination(Scheme scheme)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(scheme);
        var result = NativeMethods.SurfaceExchangeBindDestination(Handle, scheme.Handle, out var lease, out var transaction);
        if (result != GoldyResult.Ok)
            throw GoldyException.FromLastError("SurfaceExchange bind_destination");
        if (lease == nint.Zero || transaction == nint.Zero)
            throw GoldyException.FromLastError("SurfaceExchange bind_destination");
        return (new PresentLease(lease), new Transaction(transaction));
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            NativeMethods.SurfaceExchangeDestroy(Handle);
            _disposed = true;
        }
    }
}

File: ffi-client/src/surface_exchange.rs
use crate::context::Context;
use crate::error::{check, non_null_expect, Result};
use crate::exchange::Transaction;
use crate::scheme::{PresentLease, Scheme, SchemeRenderTargetLease};
use crate::sys::{self, GoldySurfaceExchange};
use crate::texture::Texture;
use crate::types::TextureFormat;

/// Window-surface exchange for present-on-scheme.
pub struct SurfaceExchange {
    ptr: *mut GoldySurfaceExchange,
}

impl SurfaceExchange {
    #[cfg(windows)]
    pub fn from_win32(ctx: &Context, hwnd: *mut std::ffi::c_void, depth: u32) -> Result<Self> {
        let ptr = non_null_expect(unsafe { sys::goldy_surface_exchange_create_win32(ctx.as_ptr(), hwnd, depth) });
        Ok(Self { ptr })
    }

    #[cfg(target_os = "macos")]
    pub fn from_appkit(ctx: &Context, ns_view: *mut std::ffi::c_void, depth: u32) -> Result<Self> {
        let ptr = non_null_expect(unsafe { sys::goldy_surface_exchange_create_appkit(ctx.as_ptr(), ns_view, depth) });
        Ok(Self { ptr })
    }

    #[cfg(target_os = "linux")]
    pub fn from_wayland(
        ctx: &Context,
        display: *mut std::ffi::c_void,
        surface: *mut std::ffi::c_void,
        depth: u32,
    ) -> Result<Self> {
        let ptr = non_null_expect(unsafe {
            sys::goldy_surface_exchange_create_wayland(ctx.as_ptr(), display, surface, depth)
        });
        Ok(Self { ptr })
    }

    pub fn size(&self) -> (u32, u32) {
        (unsafe { sys::goldy_surface_exchange_width(self.ptr) }, unsafe {
            sys::goldy_surface_exchange_height(self.ptr)
        })
    }

    pub fn format(&self) -> TextureFormat {
        unsafe { sys::goldy_surface_exchange_format(self.ptr).into() }
    }

    pub fn generation(&self) -> u64 {
        unsafe { sys::goldy_surface_exchange_generation(self.ptr) }
    }

    pub fn resize(&mut self, width: u32, height: u32) -> Result<()> {
        check(unsafe { sys::goldy_surface_exchange_resize(self.ptr, width, height) })
    }

    pub fn lease(&self) -> Result<PresentLease> {
        let ptr = non_null_expect(unsafe { sys::goldy_surface_exchange_lease(self.ptr) });
        Ok(PresentLease::from_ptr(ptr))
    }

    pub fn bind_render_target(&self, scheme: &mut Scheme, source: &SchemeRenderTargetLease) -> Result<Transaction> {
        let ptr = non_null_expect(unsafe {
            sys::goldy_surface_exchange_bind_render_target(self.ptr, scheme.as_ptr(), source.as_ptr())
        });
        Ok(Transaction::from_ptr(ptr))
    }

    pub fn bind(&self, scheme: &mut Scheme, source: &Texture) -> Result<Transaction> {
        let ptr =
            non_null_expect(unsafe { sys::goldy_surface_exchange_bind(self.ptr, scheme.as_ptr(), source.as_ptr()) });
        Ok(Transaction::from_ptr(ptr))
    }

    pub fn bind_destination(&self, scheme: &mut Scheme) -> Result<(PresentLease, Transaction)> {
        let mut out = sys::GoldySurfaceExchangeBindDestinationOut {
            lease: std::ptr::null_mut(),
            transaction: std::ptr::null_mut(),
        };
        check(unsafe { sys::goldy_surface_exchange_bind_destination(self.ptr, scheme.as_ptr(), &mut out) })?;
        Ok((
            PresentLease::from_ptr(out.lease),
            Transaction::from_ptr(out.transaction),
        ))
    }
}

impl Drop for SurfaceExchange {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { sys::goldy_surface_exchange_destroy(self.ptr) };
            self.ptr = std::ptr::null_mut();
        }
    }
}
